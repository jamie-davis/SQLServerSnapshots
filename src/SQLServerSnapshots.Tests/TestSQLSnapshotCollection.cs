using System;
using System.Collections.Generic;
using System.Text;
using SnapshotTests;
using SQLServerSnapshots.Schemas;
using SQLServerSnapshots.Snapshots;
using SQLServerSnapshots.Tests.Configuration;
using TestConsoleLib;
using TestConsoleLib.Testing;
using Xunit;

namespace SQLServerSnapshots.Tests
{
    public class TestSQLSnapshotCollection
    {
        private const string CreateRelatedTables = @"
CREATE SCHEMA Test;
GO
CREATE TABLE [Test].[A_Main]
(
    MainId INT PRIMARY KEY IDENTITY,
    Name VARCHAR(100) NULL,
    CreatedDate DATETIME NOT NULL DEFAULT(GETDATE())
);
GO
CREATE TABLE [Test].[B_Related]
(
    BId INT PRIMARY KEY IDENTITY,
    Name VARCHAR(100) NULL,
    MainId INT NOT NULL,
CONSTRAINT fk_B_MainId
    FOREIGN KEY (MainId)
    REFERENCES [Test].[A_Main] (MainId)
);
GO
";
    
        #region Types configuring test schema overrides

        [SnapshotDefinition("[Test].[AMain]")]
        public static class AMain
        {
            [Predictable]
            public static int MainId { get; set; }
        }

        #endregion
        
        public TestSQLSnapshotCollection()
        {
            DbController.TearDown();
            DbController.Run(CreateRelatedTables);
        }

        [Fact]
        public void SchemaOverridesCanBeLoadedFromType()
        {
            //Arrange
            var collection = new SqlSnapshotCollection();
            collection.ConfigureSchema(DbController.Server, DbController.TestDbName, "Test");

            //Act
            collection.LoadSchemaOverrides(GetType());

            //Assert
            var output = new Output();
            collection.GetSchemaReport(output, true);
            output.Report.Verify();
        }

        [Fact]
        public void SchemaOverridesCanBeLoadedFromAssembly()
        {
            //Arrange
            var collection = new SqlSnapshotCollection();
            collection.ConfigureSchema(DbController.Server, DbController.TestDbName, "Test");

            //Act
            collection.LoadSchemaOverrides(GetType().Assembly);

            //Assert
            var output = new Output();
            collection.GetSchemaReport(output, true);
            output.Report.Verify();
        }

        [Fact]
        public void SchemaOverridesCanBeAppliedDirectly()
        {
            //Arrange
            var collection = new SqlSnapshotCollection();
            collection.ConfigureSchema(DbController.Server, DbController.TestDbName, "Test");

            //Act
            collection.DefineTable("[Test].[A_Main]").IsPredictable("MainId");

            //Assert
            var output = new Output();
            collection.GetSchemaReport(output, true);
            output.Report.Verify();
        }
    }
}
