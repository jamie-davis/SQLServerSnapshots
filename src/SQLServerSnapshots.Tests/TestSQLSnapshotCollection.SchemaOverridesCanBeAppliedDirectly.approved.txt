Table Definitions                                                                              
-----------------                                                                              
Table Name                                                                                     
------------------                                                                             
[Test].[A_Main]                                                                                
                                                                                               
           Is      Is                                                    Referenced            
           Primary Compare Is            Is          Is       Referenced Property              
    Name   Key     Key     Unpredictable Predictable Required Table Name Name                  
    ------ ------- ------- ------------- ----------- -------- ---------- ----------            
    MainId True    False   False         True        False                                     
                                                                                               
Table Name                                                                                     
------------------                                                                             
[Test].[B_Related]                                                                             
                                                                                               
         Is      Is                                                    Referenced              
         Primary Compare Is            Is          Is       Referenced Property                
    Name Key     Key     Unpredictable Predictable Required Table Name Name                    
    ---- ------- ------- ------------- ----------- -------- ---------- ----------              
    BId  True    False   True          False       False                                       
                                                                                               
                                                                                               